### DISCUSSION TITLE
Should the time step wizard take into account of background velocity in calculation of CFL number?

### DISCUSSION THREAD

**Discussion by zhihua-zheng** [CliMA/Oceananigans.jl#3354] [General]
Hi,

I'm running a case similar to this [example](https://github.com/OceanBioME/OceanBioME.jl/blob/cdb79a3236461e5c32a916639b10dc8da997eae1/paper/figures/eady.jl), which has a background velocity from the thermal wind shear. However, I find this setup usually has some numerical instabilities at the beginning. I was wondering if that's related to the time step wizard's adjustment of `Δt` being too large. From the calculation of `cell_advection_timescale` here,
https://github.com/CliMA/Oceananigans.jl/blob/d0b7ec8f98c860ce49927e0a7214961d2f47fb75/src/Advection/cell_advection_timescale.jl#L13-L27
it seems like it does not use the background velocity. 

For this use case, should the time step wizard include the background velocity to calculate a new `Δt`?

**Comment by simone-silvestri** [comment]
You are right! Indeed, the CFL calculation should account for additional velocities. Seems like we have to update the `TimeStepWizard` to account for additional velocities. 

It should be possible by changing the model-specific `cell_advection_timescale` functions
https://github.com/CliMA/Oceananigans.jl/blob/d0b7ec8f98c860ce49927e0a7214961d2f47fb75/src/Models/HydrostaticFreeSurfaceModels/HydrostaticFreeSurfaceModels.jl#L61
and
https://github.com/CliMA/Oceananigans.jl/blob/d0b7ec8f98c860ce49927e0a7214961d2f47fb75/src/Models/NonhydrostaticModels/NonhydrostaticModels.jl#L51
to pass total velocities as done here
https://github.com/CliMA/Oceananigans.jl/blob/d0b7ec8f98c860ce49927e0a7214961d2f47fb75/src/Models/NonhydrostaticModels/nonhydrostatic_tendency_kernel_functions.jl#L65-L67

**Reply by glwagner** [reply]
For sure and just to clarify, this is not an issue with `TimeStepWizard`, but rather with the definition of `cell_advection_timescale(model)`, which is a function of each model individually.

I believe background velocities are only supported by `NonhydrostaticModel`, right?

**Comment by zhihua-zheng** [comment]
> I believe background velocities are only supported by `NonhydrostaticModel`, right?

Yes.

**Comment by glwagner** [comment]
@zhihua-zheng this is a bona fide issue, do you mind if I convert to an issue? Would you like to try to fix it?

**Reply by zhihua-zheng** [reply]
No problem with converting to an issue. I have tried some modifications according to @simone-silvestri 's suggestion, and it seems to work. No experience with fixing issues before, but I think it would be interesting to start doing it.

**Reply by glwagner** [reply]
All you gotta do is submit your new code to be merged into Oceananigans using a pull request!

**Comment by jagoosw** [comment]
I think all that needs to be change is this one line:
https://github.com/CliMA/Oceananigans.jl/blob/cc51ff0ee0a1326cebcbfa207ec0ed29ae768648/src/Models/NonhydrostaticModels/NonhydrostaticModels.jl#L51

which can be changed to:
```julia
cell_advection_timescale(model::NonhydrostaticModel) = cell_advection_timescale(model.grid, total_velocities(model)) 
```

since we defined `total_velocities` here:
https://github.com/CliMA/Oceananigans.jl/blob/cc51ff0ee0a1326cebcbfa207ec0ed29ae768648/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl#L224-L227

(plus `using Oceananigans.Models: total_velocities` at the start of the file).

**Reply by jagoosw** [reply]
I suppose arguably we need to include `biogeochemical_drift_velocity` but then we would have to compute the advection timescale for every tracer because we couldn't know how the drift velocity would effect the timescale

https://github.com/CliMA/Oceananigans.jl/blob/cc51ff0ee0a1326cebcbfa207ec0ed29ae768648/src/Models/NonhydrostaticModels/nonhydrostatic_tendency_kernel_functions.jl#L250-L254

**Reply by glwagner** [reply]
> (plus using Oceananigans.Models: total_velocities at the start of the file).

You won't need that because `cell_advection_timescale` is defined in the same module as `total_velocities`

**Reply by glwagner** [reply]
Another possibility is

```julia
function cell_advection_timescale(model::NonhydrostaticModel)
    grid = model.grid
    velocities = total_velocities(model)
    return cell_advection_timescale(grid, velocities)
end
```

**Reply by zhihua-zheng** [reply]
I think these are all good solutions! Just make sure I understand the source code here: Both your and @jagoosw's modification of `cell_advection_timescale` in [NonhydrostaticModels.jl](https://github.com/CliMA/Oceananigans.jl/blob/cc51ff0ee0a1326cebcbfa207ec0ed29ae768648/src/Models/NonhydrostaticModels/NonhydrostaticModels.jl#L51) is to redefine the method of `cell_advection_timescale` in `NonhydrostaticModel`. This method is being used in https://github.com/CliMA/Oceananigans.jl/blob/cc51ff0ee0a1326cebcbfa207ec0ed29ae768648/src/Simulations/time_step_wizard.jl#L102
with `model` as the input, to compute a new time step. Is this interpretation correct?


In addition to @jagoosw's comment about `biogeochemical_drift_velocity`: does the velocity in `AdvectiveForcing` also play a role here?

**Comment by zhihua-zheng** [comment]
> You won't need that because `cell_advection_timescale` is defined in the same module as `total_velocities`

Because `total_velocities` that defined in https://github.com/CliMA/Oceananigans.jl/blob/cc51ff0ee0a1326cebcbfa207ec0ed29ae768648/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl#L223-L227
is included by this line below? Sorry, I'm new to Julia and the code structure seems a bit convoluted.
https://github.com/CliMA/Oceananigans.jl/blob/cc51ff0ee0a1326cebcbfa207ec0ed29ae768648/src/Models/NonhydrostaticModels/NonhydrostaticModels.jl#L42

**Reply by glwagner** [reply]
The module I was referring to is `NonhydrostaticModels`. The top-level file `NonhydrostaticModels.jl` includes all the code for the module `NonhydrostaticModels`. The directory containing all the code for the module `NonhydrostaticModels` is called `NonhydrostaticModels`:

<img width="632" alt="image" src="https://github.com/CliMA/Oceananigans.jl/assets/15271942/858f900f-6614-46d3-abcb-2d4a5c9a7703">

We are definitely open to concrete suggestions or PRs that improve the code structure and make it less convoluted, which is highly undesirable.
